pkg/frontend/middleware/maintenance.go

```diff
@@ -0,0 +1,51 @@
+package middleware
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"net/http"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/Azure/ARO-RP/pkg/metrics"
+)
+
+type MaintenanceMiddleware struct {
+	metrics.Emitter
+}
+
+// Emit metric for unplanned maintenance
+func (mm MaintenanceMiddleware) UnplannedMaintenanceSignal(h http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx, cancel := context.WithCancel(r.Context())
+		defer cancel()
+
+		resourceID := strings.TrimPrefix(filepath.Dir(r.URL.Path), "/admin")
+
+		// Use a do-while loop to ensure we emit the metric at least once
+		mm.emitMaintenanceSignal("unplanned", resourceID)
```

Comment:
I'm wondering if we want to emit a second metric for planned maintenance as well, or just have a single maintenance signal that emits when maintenance is occurring generally (regardless of if planned / unplanned)

Reply:
I agree having one signal "source" is ideal. The main issue is unplanned maintenance only uses the synchronous, frontend operations. PUCM uses asynchronous, backend operations.